src/textual/worker.py

```diff
@@ -0,0 +1,388 @@
+
+
+class NoActiveWorker(Exception):
+    """There is no active worker."""
+
+
+class WorkerError(Exception):
+    """A worker related error."""
+
+
+class WorkerFailed(WorkerError):
+    """The worker raised an exception and did not complete."""
+
+    def __init__(self, error: BaseException) -> None:
+        self.error = error
+        super().__init__(f"Worker raise exception: {error!r}")
+
+
+class DeadlockError(WorkerError):
+    """The operation would result in a deadlock."""
+
+
+class WorkerCancelled(WorkerError):
+    """The worker was cancelled and did not complete."""
+
+
+def get_current_worker() -> Worker:
+    """Get the currently active worker.
+
+    Raises:
+        NoActiveWorker: If there is no active worker.
+
+    Returns:
+        A Worker instance.
+    """
+    try:
+        return active_worker.get()
+    except LookupError:
+        raise NoActiveWorker("There is no active worker in this task or thread.")
+
+
+class WorkerState(enum.Enum):
+    """A description of the worker's current state."""
+
+    PENDING = 1
+    """Worker is initialized, but not running."""
+    RUNNING = 2
+    """Worker is running."""
+    CANCELLED = 3
+    """Worker is not running, and was cancelled."""
+    ERROR = 4
+    """Worker is not running, and exited with an error."""
+    SUCCESS = 5
+    """Worker is not running, and completed successfully."""
+
+
+ResultType = TypeVar("ResultType")
+
+
+WorkType: TypeAlias = Union[
+    Callable[[], Coroutine[None, None, ResultType]],
+    Callable[[], ResultType],
+    Awaitable[ResultType],
+]
+
+
+class _ReprText:
+    """Shim to insert a word in to the Worker's repr."""
+
+    def __init__(self, text: str) -> None:
+        self.text = text
+
+    def __repr__(self) -> str:
+        return self.text
+
+
+@rich.repr.auto(angular=True)
+class Worker(Generic[ResultType]):
+    """A class to manage concurrent work (either a task or a thread)."""
+
+    @rich.repr.auto
+    class StateChanged(Message, bubble=False):
+        """The worker state changed."""
+
+        namespace = "worker"
+
+        def __init__(self, worker: Worker, state: WorkerState) -> None:
+            """Initialize the StateChanged message.
+
+            Args:
+                worker: The worker object.
+                state: New state.
+            """
+            self.worker = worker
+            self.state = state
+            super().__init__()
+
+        def __rich_repr__(self) -> rich.repr.Result:
+            yield self.worker
+            yield self.state
+
+    def __init__(
+        self,
+        node: DOMNode,
+        work: WorkType | None = None,
+        *,
+        name: str = "",
+        group: str = "default",
+        description: str = "",
+        exit_on_error: bool = True,
+    ) -> None:
+        """Initialize a Worker.
+
+        Args:
+            node: THe widget, screen, or App that initiated the work.
+            work: A callable, coroutine, or other awaitable.
```

Comment:
Instead of the types, wouldn't it make more sense that `work` is supposed to be an object that holds the work / does the work / makes things happen?
I.e. the thing that will actually "run" and update the app?

Reply:
I don't follow what you are suggesting. Can you elaborate?

Reply:
A word was eaten out of my comment.
"Wouldn't it make more sense to _write_ that `work` is ..."
The current docstring lists the valid types for `work` but it doesn't say what that callable/coroutine/awaitable is used for.